### PR TITLE
Add module MID

### PR DESCRIPTION
**Pull request by xa1lopez**

### CODE REVIEW

File: Modules/MID/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# ---- Library ----
+
+add_library(QcMIDRaw)
```
Comment (Barthelemy) [inline Modules/MID/CMakeLists.txt:3]: I would propose to stick to a single library per detector, thus `QcMID` and to have all the classes in there (raw, cluster, you name it). I know that ITS had a raw library but it was the beginning and we did not know better.

File: Modules/MID/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+add_root_dictionary(QcMIDRaw
+                    HEADERS
+		    include/MID/RawQcCheck.h
+		    include/MID/RawQcTask.h
+                    LINKDEF include/MID/LinkDef.h
+                    BASENAME QcMIDRaw)
+
+# ---- Test(s) ----
+
+#set(TEST_SRCS test/testQcMID.cxx)
```
Comment (Barthelemy) [inline Modules/MID/CMakeLists.txt:29]: I can merge without it, but please consider adding some kind of tests later on.

File: Modules/MID/include/MID/RawQcCheck.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+///
+/// \file   RawQcCheck.h
+/// \author My Name
```
Comment (Barthelemy) [inline Modules/MID/include/MID/RawQcCheck.h:13]: My name -> Xavier Lopez (or someone else)

File: Modules/MID/include/MID/RawQcCheck.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+#ifndef QC_MODULE_MID_MIDRAWQCCHECK_H
+#define QC_MODULE_MID_MIDRAWQCCHECK_H
+
+#include "QualityControl/CheckInterface.h"
+
+namespace o2::quality_control_modules::mid
+{
+
+/// \brief  Example QC Check
```
Comment (Barthelemy) [inline Modules/MID/include/MID/RawQcCheck.h:24]: update doxygen. Does not need to be long, just a few words to say what it is and what it does. I put this comment here but it applies to the Task as well.

File: Modules/MID/src/RawQcCheck.cxx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    } // end check mDetElemID
+  }
+  return result;
+}
+
+std::string RawQcCheck::getAcceptedType() { return "TH1"; }
+
+void RawQcCheck::beautify(std::shared_ptr<MonitorObject> mo, Quality checkResult)
+{
+  if (mo->getName() == "example") {
```
Comment (Barthelemy) [inline Modules/MID/src/RawQcCheck.cxx:80]: you can probably remove this code (but keep the empty method).

File: Modules/MID/raw-mid.json [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        "detectorName": "MID",
+        "cycleDurationSeconds": "10",
+        "maxNumberCycles": "-1",
+        "dataSource_comment": "The other type of dataSource is \"direct\", see basic-no-sampling.json.",
+        "dataSource": {
+          "type": "dataSamplingPolicy",
+          "name": "digits"
+        },
+        "taskParameters": {
+          "myOwnKey": "myOwnValue"
```
Comment (Barthelemy) [inline Modules/MID/raw-mid.json:39]: probably not needed ?

File: Modules/MID/src/RawQcTask.cxx [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    delete mDetElemID;
+  }
+}
+
+void RawQcTask::initialize(o2::framework::InitContext& /*ctx*/)
+{
+  ILOG(Info) << "initialize RawQcTask" << ENDM; // QcInfoLogger is used. FairMQ logs will go to there as well.
+
+  // this is how to get access to custom parameters defined in the config file at qc.tasks.<task_name>.taskParameters
+  if (auto param = mCustomParameters.find("myOwnKey"); param != mCustomParameters.end()) {
```
Comment (Barthelemy) [inline Modules/MID/src/RawQcTask.cxx:40]: this piece is probably not needed.

File: Modules/MID/src/RawQcTask.cxx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    ILOG(Info) << "Custom parameter - myOwnKey: " << param->second << ENDM;
+  }
+
+  mDetElemID = new TH1F("mDetElemID", "Id of detector element", 81, -0.5, 80.5);
+  getObjectsManager()->startPublishing(mDetElemID);
+  try {
+    getObjectsManager()->addMetadata(mDetElemID->GetName(), "custom", "34");
```
Comment (Barthelemy) [inline Modules/MID/src/RawQcTask.cxx:47]: probably not needed ?

File: Modules/MID/src/runMID.cxx [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+{
+  DataSampling::CustomizeInfrastructure(policies);
+}
+
+void customize(std::vector<ConfigParamSpec>& workflowOptions)
+{
+  workflowOptions.push_back(
+    ConfigParamSpec{ "config-path", VariantType::String, "", { "Path to the config file. Overwrite the default paths. Do not use with no-data-sampling." } });
+  workflowOptions.push_back(
+    ConfigParamSpec{ "no-data-sampling", VariantType::Bool, false, { "Skips data sampling, connects directly the task to the producer." } });
```
Comment (Barthelemy) [inline Modules/MID/src/runMID.cxx:39]: I don't think that you use this parameter

**Review by Barthelemy — COMMENTED** [review]
Thank you for your contribution. I have put a few comments. All is good in general, just a few remnant code from the examples.

**Comment by Barthelemy** [comment]
if you need clarification or help, let me know

**Comment by xa1lopez** [comment]
I made most of the fixes that you suggest, thank's

**Comment by Barthelemy** [comment]
Hi, thank you. 
I think that there is a problem with the rebase. Could you try to clean up my commits ? 
If you need help let me know. 
Cheers,

**Comment by xa1lopez** [comment]
I rebased, hope it's ok

**Comment by xa1lopez** [comment]
It looks like i have still a problem with my Clang version, i guess i need your help for this (cc7 with the jira ticket https://alice.its.cern.ch/jira/browse/O2-1600). In the mean time i will try to solve this by saving your fix formatting and make a diff with my own formatting while running my clang version. Sorry for this.

**Comment by xa1lopez** [comment]
I manage Clang please ignore my last comment

**Comment by Barthelemy** [comment]
great !

**Comment by Barthelemy** [comment]
waiting for the tests.

**Comment by xa1lopez** [comment]
Dear Barth,
sorry to spam you, is it normal that those checks take 8 days ? Should i do something ?
Many thank's,
Xavier

**Comment by Barthelemy** [comment]
Dear Xavier, no worries, you are not spamming at all. It is absolutely not normal. It should take a couple of hours. I did not noticed becase I check only the new or green PRs. I will try to close and reopen the PR.

**Comment by Barthelemy** [comment]
@ktf any idea what is going on here ? why are the tests stuck ?

**Comment by ktf** [comment]
"First-time contributor". Needs to have an approved review, or to be added to the list of members. Explained at:

https://aliceo2group.github.io/quickstart/devel.html#pull-request-flow

**Comment by Barthelemy** [comment]
forgot about that, once again. Thanks @ktf. I will review first then.

**Comment by Barthelemy** [comment]
actually I had review, ok rereviewing

**Review by Barthelemy — APPROVED** [review]

**Comment by ktf** [comment]
I think it want an "approved" review. If you approve it once and then push another commit, it will still consider it as "first timer".

**Comment by Barthelemy** [comment]
sorry for the delay

**Comment by xa1lopez** [comment]
Hip Hip Houra ! no sorry at all and thank you for your commitment and your prompt reaction !

**Pull request merged**